### PR TITLE
powerloader refactor: load anything

### PR DESCRIPTION
**Pull request by Fikou**
## About The Pull Request

adds a powerloader connected component that keeps track of a powerloader
removes keeping track of a powerloader in powerloader clamps, you can now use them wherever you want if an admin spawns them
refactors the loading things that are called like 20 times to use a proc since they are all the same
makes ripleys actually use batteries because theres a reason they exist right??
adds the screwpley its a powerloader with a screwdriver and wrench to showcase the possibilities
adds a custom powerloader, an admin can select any item they want and put it as one of the powerloaders hands, so you could have a katana shield powerloader if you wanted it
video: https://streamable.com/mshytv

## Why It's Good For The Game

fun

## Changelog
:cl:
code: improves powerloader code
admin: you can now create custom powerloaders
/:cl:

### CODE REVIEW

**Comment by ilikesaying** [comment]
one step closer to combat mechs

File: code/modules/vehicles/powerloader.dm [hunk 1]
```diff
@@ -301,16 +255,66 @@
...
+
+/obj/vehicle/powerloader/custom/attack_ghost(mob/user)
+	if(!IsAdminGhost(user) || setup)
+		return
+	var/chosen_one
+	var/chosen_two
+	for(var/i in 1 to 2)
+		var/object = input(user, "What to add?", text("Input")) as text
+		if(!object || object == "")
+			continue
```
Comment (TiviPlus) [inline code/modules/vehicles/powerloader.dm:308]: Empty strings are falsy

File: code/modules/vehicles/powerloader.dm [hunk 2]
```diff
@@ -301,16 +255,66 @@
...
+		var/chosen = pick_closest_path(object)
+		if(!chosen)
+			continue
+		if(!ispath(chosen, /obj/item))
+			to_chat(user, "<span class='warning'>[chosen] is not an item, please try again.</span>")
```
Comment (TiviPlus) [inline code/modules/vehicles/powerloader.dm:313]: wouldnt null fail ispath anyway

File: code/modules/vehicles/powerloader.dm [hunk 3]
```diff
@@ -301,16 +255,66 @@
 	else
 		icon_state = "loader_clamp"
 
-/obj/item/powerloader_clamp/attack_self(mob/user)
-	if(user in linked_powerloader.buckled_mobs)
-		linked_powerloader.unbuckle_mob(user)
+/obj/item/powerloader_clamp/proc/load(atom/movable/target, mob/user)
```
Comment (TiviPlus) [inline code/modules/vehicles/powerloader.dm:258]: Autodoc

File: code/modules/vehicles/powerloader.dm [hunk 4]
```diff
@@ -237,59 +218,32 @@
...
-		RD.forceMove(linked_powerloader)
-		loaded = RD
-		playsound(src, 'sound/machines/hydraulics_2.ogg', 40, 1)
-		update_icon()
-		user.visible_message("<span class='notice'>[user] grabs [loaded] with [src].</span>",
-		"<span class='notice'>You grab [loaded] with [src].</span>")
+		load(RD, user)
 
 	else if(istype(target, /obj/structure/ore_box))
 		var/obj/structure/ore_box/OB = target
```
Comment (TiviPlus) [inline code/modules/vehicles/powerloader.dm:245]: Uneeeded cast

File: code/modules/vehicles/powerloader.dm [hunk 5]
```diff
@@ -237,59 +218,32 @@
...
-		playsound(src, 'sound/machines/hydraulics_2.ogg', 40, 1)
-		update_icon()
-		user.visible_message("<span class='notice'>[user] grabs [loaded] with [src].</span>",
-		"<span class='notice'>You grab [loaded] with [src].</span>")
+		load(V, user)
 
 	else if(istype(target, /obj/structure/reagent_dispensers))
 		var/obj/structure/reagent_dispensers/RD = target
 		if(RD.anchored)
 			to_chat(user, "<span class='warning'>You can't lift this!</span>")
```
Comment (TiviPlus) [inline code/modules/vehicles/powerloader.dm:240]: You can condense some of this elseif yandere code with ||

File: code/modules/vehicles/powerloader.dm [hunk 6]
```diff
@@ -95,28 +88,30 @@
...
+	icon_state = background_icon
+	overlays += image(icon_state = overlay_icon, layer = MOB_LAYER + 0.1)
 	move_delay = max(4, move_delay - buckling_mob.skills.getRating("powerloader"))
-	var/clamp_equipped = 0
-	for(var/obj/item/powerloader_clamp/PC in contents)
-		if(!buckling_mob.put_in_hands(PC))
-			PC.forceMove(src)
+	var/equipment_equipped = 0
+	var/equipment_required = 0
+	for(var/obj/item/hand in list(hand_one, hand_two))
```
Comment (TiviPlus) [inline code/modules/vehicles/powerloader.dm:96]: Why the typecheck in this list

File: code/modules/vehicles/powerloader.dm [hunk 7]
```diff
@@ -95,28 +88,30 @@
 /obj/vehicle/powerloader/post_buckle_mob(mob/buckling_mob)
 	. = ..()
 	playsound(loc, 'sound/mecha/powerloader_buckle.ogg', 25)
-	icon_state = "powerloader"
-	overlays += image(icon_state= "powerloader_overlay", layer = MOB_LAYER + 0.1)
+	icon_state = background_icon
+	overlays += image(icon_state = overlay_icon, layer = MOB_LAYER + 0.1)
```
Comment (TiviPlus) [inline code/modules/vehicles/powerloader.dm:92]: managed overlays?

File: code/modules/vehicles/powerloader.dm [hunk 8]
```diff
@@ -39,26 +43,15 @@
...
-
-	if(istype(I, /obj/item/powerloader_clamp))
-		var/obj/item/powerloader_clamp/PC = I
-		if(PC.linked_powerloader != src)
-			return
-
-		return user_unbuckle_mob(user, user) //clicking the powerloader with its own clamp unbuckles the pilot.
-
-	else if(isscrewdriver(I))
+	if(isscrewdriver(I))
```
Comment (TiviPlus) [inline code/modules/vehicles/powerloader.dm:46]: Screwdriver act

File: code/game/objects/structures/orbital_cannon.dm [hunk 9]
```diff
@@ -329,20 +324,11 @@
...
 			return TRUE
-
-		forceMove(PC.linked_powerloader)
-		PC.loaded = src
-		playsound(loc, 'sound/machines/hydraulics_2.ogg', 40, 1)
-		PC.update_icon()
-		to_chat(user, "<span class='notice'>You grab [PC.loaded] with [PC].</span>")
+		PC.load(src, user)
 		update_icon()
 		return TRUE
```
Comment (TiviPlus) [inline code/game/objects/structures/orbital_cannon.dm:333]: Nonblocking, but this should be with the yandere elseifs

File: code/game/objects/structures/dropship_ammo.dm [hunk 10]
```diff
@@ -25,39 +25,33 @@
...
+						to_chat(user, "<span class='notice'>You transfer [transf_amt] [ammo_name] to [src].</span>")
+						if(!SA.ammo_count)
+							PC.loaded = null
+							PC.update_icon()
+							qdel(SA)
+		else
+			PC.load(src, user)
+			update_icon()
+		return TRUE
+	. = ..()
```
Comment (TiviPlus) [inline code/game/objects/structures/dropship_ammo.dm:49]: Hello yes early returns

**Review by TiviPlus — CHANGES_REQUESTED** [review]

File: code/modules/vehicles/powerloader.dm (hunk 6 above)
Comment (Fikou) [reply code/modules/vehicles/powerloader.dm:96]: why not

Comment (TiviPlus) [reply code/modules/vehicles/powerloader.dm:96]: Because it's slower and implies the list isn't a type

**Pull request closed without merging**